src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -599,6 +641,111 @@ private static ImmutableArray<RefKind> GetRefKindsOrNull(ArrayBuilder<RefKind> r
             }
         }
 
+        // This fills in the arguments and parameters arrays in evaluation order.
+        private static ImmutableArray<IArgument> BuildArgumentsInEvaluationOrder(
+            SyntaxNode syntax,
+            Symbol methodOrIndexer,
+            MethodSymbol optionalParametersMethod,
+            bool expanded,
+            ImmutableArray<int> argsToParamsOpt,
+            ImmutableArray<BoundExpression> arguments,
+            Binder binder)
+        {
+            ImmutableArray<ParameterSymbol> parameters = methodOrIndexer.GetParameters();
+
+            ArrayBuilder<IArgument> argumentsInEvaluationBuilder = ArrayBuilder<IArgument>.GetInstance(parameters.Length);
+
+            PooledHashSet<int> processedParameters = PooledHashSet<int>.GetInstance();
+
+            // First, fill in all the explicitly provided arguments.
+            for (int a = 0; a < arguments.Length; ++a)
+            {
+                BoundExpression argument = arguments[a];
+
+                int p = (!argsToParamsOpt.IsDefault) ? argsToParamsOpt[a] : a;
+                var parameter = parameters[p];
+
+                Debug.Assert(!processedParameters.Contains(p));
+
+                processedParameters.Add(p);
+
+                ArgumentKind kind = ArgumentKind.Explicit;
+
+                if (IsBeginningOfParamArray(p, a, expanded, parameters.Length, arguments, argsToParamsOpt, out int paramArrayArgumentCount))
+                {
+                    int firstNonParamArrayArgumentIndex = a + paramArrayArgumentCount;
+                    Debug.Assert(firstNonParamArrayArgumentIndex <= arguments.Length);
+
+                    kind = ArgumentKind.ParamArray;
+                    ArrayBuilder<BoundExpression> paramArray = ArrayBuilder<BoundExpression>.GetInstance(paramArrayArgumentCount);
+
+                    for (int i = a; i < firstNonParamArrayArgumentIndex; ++i)
+                    {
+                        paramArray.Add(arguments[i]);
+                    }
+
+                    // Set loop variable so the value for next iteration will be the index of the first non param-array argument after param-array argument(s).
+                    a = firstNonParamArrayArgumentIndex - 1;
+
+                    argument = CreateParamArrayArgument(syntax, parameter.Type, paramArray.ToImmutableAndFree(), null, binder);
+                }
+
+                argumentsInEvaluationBuilder.Add(BoundCall.CreateArgumentOperation(kind, parameter, argument)); 
+            }    
+
+            // Collect parameters with missing arguments.   
+            ArrayBuilder<ParameterSymbol> missingParametersBuilder = ArrayBuilder<ParameterSymbol>.GetInstance(parameters.Length);
+            for (int i = 0; i < parameters.Length; ++i)
+            {
+                if (!processedParameters.Contains(i))
+                {
+                    missingParametersBuilder.Add(parameters[i]);
```

Comment:
>missingParametersBuilder.Add(parameters[i]); [](start = 20, length = 44)

I think we can simplify this code by changing AppendMissingOptionalArguments to handle just one parameter at a time (basically by removing the loop from it, but leaving the loop body) and calling it from here. It could simply return IArgument. Then, we don't need to fill missingParametersBuilder and pass argumentsInEvaluationBuilder in. It is Ok to follow-up on this in a separate PR, I think.